lib/nodes/fpga.cpp

```diff
@@ -168,19 +172,93 @@ const std::string &FpgaNode::getDetails() {
 int FpgaNode::check() { return 0; }
 
 int FpgaNode::start() {
-  // enque first read
-  // dma->read(*(blockRx[0]), blockRx[0]->getSize());
+  if (getInputSignalsMaxCount() * sizeof(float) > blockRx->getSize()) {
+    logger->error("Input signals exceed block size.");
+    throw villas ::RuntimeError("Input signals exceed block size.");
+  }
+  if (lowLatencyMode) {
+    dma->readScatterGatherPrepare(*blockRx, blockRx->getSize());
```

Comment:
@stv0g can I get your opinion on this?

Basically, when lowLatencyMode is true, I setup the read and write transactions in `start`. This means I only need to start the DMA controller in `read` and `write`. However, we cannot change the write length and do not retrieve only minimal status information from the DMA controller.

From the commit message:
This setting improves latency by remove various checks.
Use with caution! Requires read cache in FPGA design!
The common use case in VILLASfpga is that we have exactly
one write for every read and the number of exchanged signals
do not change. If this is the case, we can reuse the buffer
descriptors during reads and write, thus avoidng freeing,
reallocating and setting them up.
We set up the descriptors in start, and in write or read,
we only reset the complete bit in the buffer descriptor and
write to the tdesc register to start the DMA transfer.
Improves read/write latency by approx. 40%.

Reply:
Is it correct to assume that getInputSignalsMaxCount is the number of signals passed to write and read?

Reply:
For read this can be assumed as the node itself defines via its `in = { signals = (..) }` configuration which and how many signals should be read. This (currently) can not change during runtime and hence you can initialize this already in `start()`.

But I think we allow partial reads. The node might decide to provide less then the maximum number of signals per read. So this should be understood as an upper bound.

Thats also true for `write()`. If a sourcing node provide less signals than configured during its `read()`, we can have situations where the corresponding `write()` of the forewarded data sees less than the maximum number of signals.

Reply:
This implementation would allow partial reads, but requires fixed length writes. If there are less signals than set during init, fastWrite would reuse the data from the previous timestep.

I documented this in a comment in the code.